### ISSUE TITLE
Support ES6 features

### ISSUE THREAD

**Issue by kevinprotoss** [caolan/async#719]
Hello,

I found `async` module does not support some es6 features such as context binding while using arrow function.

Is it possible to support this in the future?

Best regards,
Kevin

**Comment by beaugunderson** [comment]
Do you have an example of what doesn't work? That would be very helpful, thanks!

---

**Comment by kevinprotoss** [comment]
Hello,

In ES 6: 
`Unlike functions, arrows share the same lexical this as their surrounding code.`

This means if I have some codes like the following

```js
var bob = {
  _name: "Bob",
  _friends: [],
  printFriends() {
    this._friends.forEach(f =>
      console.log(this._name + " knows " + f));
  }
}
```

However if I use the async module by means of arrows, I can't get the same lexical in the callback of async
e.g.,
```js
var bob = {
  _name: "Bob",
  _friends: [],
  printFriends: function() {
    async.each(this._friends, (f, ascb) => {
      console.log(this._name + " knows " + f));
      ascb();
    }, () => {
      console.log("finish printing all friends of " + this._name);
    });
  }
}
```

Best regards,
Kevin

**Comment by aearly** [comment]
You can accomplish the same goal today in ES5 with `bind`:

```js
var bob = {
  _name: "Bob",
  _friends: [],
  printFriends: function() {
    async.each(this._friends, function (f, ascb) {
      console.log(this._name + " knows " + f));
      ascb();
    }.bind(this), function () {
      console.log("finish printing all friends of " + this._name);
    }.bind(this);
  }
}
```

It's a bit ugly, but that's why people  recommend avoiding `this` in async code --  just use explicit args.

I would love to support ES6 features -- there are tons of handy things (spread operators, default arguments) -- but seeing as async is still widely used in node 0.10 and in the browser, it would force a lot of people to use a transpiler.

**Comment by khanh-nguyen** [comment]
How about something like this?
```
async.eachSeries(arr, function*(item) {
    yield doSomething(item);
    console.log('Did something with', item);
}, function(err) {
    if (err) console.log(err);
    else console.log('All good.');
});
```
Is it supported? I tried this but didn't get any console.log message at all.

**Comment by kevinprotoss** [comment]
@aearly I agree with you. I also know the method by using ES5 but I don't like that. I think lots of ES6 features are useful. For example, If I want to use async module in a generator, I had to use a thunk for wrapping the async. Is it possible to let some async functions return generator?

**Comment by khanh-nguyen** [comment]
Never mind, that was a bad example. With yeild, I don't need `eachSeries` anyway.

**Comment by aearly** [comment]
async is narrowly focused on working with node-flavored continuation passing style functions.  If you want to use generators with async, or async with generators, you'll have to wrap things top make it work -- just like using node's built-in functions with generators.

(I actually would say that ES6 generators make async a bit obsolete -- you can do some of the things async does more nicely with generators, but there still are many people supporting ES5 environments with async.)

Going back to your first question/example, you can use arrow functions just fine without any support from async (check the console):

http://codepen.io/anon/pen/jEvMeg

There was a syntax error in the code you provided.

**Comment by AntJanus** [comment]
There seems to be another issue other than just generators. I used `async.waterfall` like so:

````
//console.log(this) would result in an object listing all of my class variables
async.waterfall([
  this.classFunction,
  this.classFunction2,
  this.classFunction3
]);
````

I used it inside a class where `this.classFunction` refers to a class function and got some strange behavior. While the functions would run, their `this` did not refer to the class `this` but rather to a null value. So, if my `classFunction` definition was like so:

`````
classFunction(cb) {
  //console.log(this) results in null
  cb(null, 'some value');
}
`````

The inside of the class function would no longer have an access to the class's `this` but only if run through any `async` function. The only way to prevent that is to litter the code with `.bind(this)` like so:

````
async.waterfall([
  this.classFunction.bind(this),
  this.classFunction2.bind(this),
  this.classFunction3.bind(this)
]);
````

And then the `classFunctions` have access to the class variables.
